Objects/cown.c

```diff
@@ -0,0 +1,323 @@
+        if (o->owning_thread != tstate->thread_id) { \
+            PyErr_Format(PyExc_RegionError, "%s: %S", msg, o); \
+            return NULL; \
+        } \
+    } while(0);
+
+#define bail_unless_in_state(o, expected_state, msg) \
+    do { \
+        /* Note: we must hold the GIL at this point -- note for future threading implementation. */ \
+        if (o->state != expected_state) { \
+            PyErr_Format(PyExc_RegionError, "%s: %S", msg, o); \
+            return NULL; \
+        } \
+    } while(0);
+
+#define bail_unless_acquired(o, msg) \
+    bail_unless_owned(o, msg) \
+    bail_unless_in_state(o, Cown_ACQUIRED, msg)
+
+static PyObject *PyCown_acquire(PyCownObject *self, PyObject *args) {
+    int expected = 0;
+
+    while (!atomic_compare_exchange_strong(&self->state, &expected, Cown_ACQUIRED)) {
+        sem_wait(&self->semaphore);
+        expected = 0;
+    }
+
+    // Note: we must hold the GIL at this point -- note for future
+    // threading implementation.
+    PyThreadState *tstate = PyThreadState_Get();
+    self->owning_thread = tstate->thread_id;
+
+#ifndef NDEBUG
+    fprintf(stderr, "Cown %p acquired by %zd\n", self, tstate->thread_id);
+#endif
+    Py_RETURN_NONE;
+}
+
+static PyObject *PyCown_release(PyCownObject *self, PyObject *args) {
+    if (self->state == Cown_RELEASED) {
+        bail_if_owned(self, "BUG: Released cown had owning thread: %p");
+        Py_RETURN_NONE;
+    }
+
+    bail_unless_owned(self, "Thread attempted to release a cown it did not own");
+
+    atomic_store(&self->state, Cown_RELEASED);
+    self->owning_thread = 0;
+    sem_post(&self->semaphore);
+
+#ifndef NDEBUG
+    fprintf(stderr, "Cown %p released\n", self);
+#endif
+    Py_RETURN_NONE;
+}
+
+int _PyCown_release(PyObject *self) {
+    PyObject* res = PyCown_release((PyCownObject *)self, NULL);
+    return res == Py_None ? 0 : -1;
+}
+
+int _PyCown_is_released(PyObject *self) {
+    PyCownObject *cown = (PyCownObject *)self;
+    return cown->state == Cown_RELEASED;
+}
+
+static PyObject *PyCown_get(PyCownObject *self, PyObject *args) {
+    bail_unless_acquired(self, "Attempt to get value of unacquired cown");
+
+    if (self->value) {
+        return Py_NewRef(self->value);
+    } else {
+        Py_RETURN_NONE;
+    }
+}
+
+// Needed to test for region object
+extern PyTypeObject PyRegion_Type;
+extern PyTypeObject PyCown_Type;
+
+static PyObject *PyCown_set_unchecked(PyCownObject *self, PyObject *args) {
+    // Cowns are cells that hold a reference to a bridge object,
+    // (or another cown or immutable object)
+    const bool is_region_object =
+        args->ob_type == &PyRegion_Type && _Py_is_bridge_object(args);
+    if (is_region_object ||
+        args->ob_type == &PyCown_Type ||
+        _Py_IsImmutable(args)) {
+        // Pyrona_ADDREFERENCE(_PyObject_CAST(self), args);
+
+        PyObject* old = self->value;
+        self->value = args;
+        Py_XINCREF(self->value);
+        Py_XDECREF(old);
```

Comment:
I think this is something that we should look for elsewhere. But not as part of this PR.  Just something to all pay attention to.
```suggestion
        Py_XINCREF(args);
        Py_XSETREF(self->value, args);
```